resources/lambdas.md

```diff
@@ -78,23 +78,64 @@ For more on lambda expressions, refer to the
+While this may seem useless, it is powerful to define incredibly complex processes, and only commit processing power when it is needed (lazy evaluation).
+
+
+```text
+f(x) = y
+
+-- snip --
+
+f(x) -> [what happens in f(x) stays in f(x)...] -> y
+```
+
+From these implementations, two things are realized; namely:
+- Working with sets becomes very useful
+- Functions can't *do anything* besides define more things. (A lot like mathematical functions)
+
+What does that last bit mean? Basically, a function cannot contain [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). In other words, it must follow a
+"black box" model, where all the function/caller can see is the value passed in and the value returned, nothing outside edited, no prints, no outside variables, nada.
+
+Functional programming languages are still widely used, particularly in data science for large processing operations. The YouTube channel Computerphile has a ton of 
+really neat videos on them if you're interested.
+
+## What does this have to do with what I'M working with?
+
+Great question. Well, since around 1993, a lot of imperative language developers started to realize how powerful functional programming can be.
+As you'd expect, Java is one of these functional-capable languages, and you may have even seen some of its functional features already. The most commonly used ones are:
+
+- Anonymous functions (lambdas): functions with no name
+- `map` function: applies a given function to every element in a collection
+- `fold` function: collapses a collection into a single value according to a passed function
+- `filter` function: returns a collection with elements kept/removed according to a boolean function passed
+- Closures: while java does not have support for closures, these can be emulated with anonymous classes
+
+A typical usecase of functional programming in Java might be:
 ```java
 int[] things = {1, 2, 3, 4, 5, 6, 7, 8, 9};
 int[] otherThings = Arrays.stream(things).filter(i->i>5).toArray();
 ```
+
+There is a lot to note here about the aforementioned "side effects" as well.
+
 This program created a new array that only contained the integers that were greater than 5 from `things[]`. The important detail is that `things[]` was never altered; there were no side effects. 
 As a result, streams are an example of functional programing (but not the only one) since they never alter the original collection, and lambdas are too (as long as the input never gets altered). If you like to think of it mathematically, f(x) 
 must always return the same number (let's say a); but if you change the value of x it won't return a. If x never gets changed, then we have functional
-programming; but if it doesn't then we don't.
+programming; but if it doesn't, it must be pulling from an outside source, no bueno.
```

Comment:
It's not necessarily pulling from an outside source; it just got modified due to a "side effect".